src/DNSC/Cache.hs

```diff
@@ -172,32 +172,30 @@ insertRRs now rrs rank c = insertRRSet =<< takeRRSet rrs
 @
  -}
 insert :: Timestamp -> Key -> TTL -> CRSet -> Ranking -> Cache -> Maybe Cache
-insert now k@(Key dom typ cls) ttl crs rank c =
-  maybe inserted withOldRank lookupRank
+insert now k@(Key dom typ cls) ttl crs rank cache@(Cache c xsz) =
+  maybe sized withOldRank lookupRank
   where
     lookupRank =
       lookup_ now (\_ _ (Val _ r) -> r)
-      dom typ cls c
+      dom typ cls cache
     withOldRank r = do
       guard $ rank > r
-      inserted
+      sized
     eol = now <+ ttl
-    inserted =
-      return $ PSQ.insert k eol (Val crs rank) c
+    sized
+      | PSQ.size inserted <= xsz  =  return $ Cache inserted xsz
+      | otherwise                 =  do
+          (_, _, _, deleted) <- PSQ.minView inserted
+          return $ Cache deleted xsz
+    inserted = PSQ.insert k eol (Val crs rank) c
 
-expires :: Timestamp -> Cache -> Maybe Cache
-expires now = rec0
-  where
-    rec0 c = rec1 <$> expire1 now c
-    rec1 c = maybe c rec1 $ expire1 now c
 
-expire1 :: Timestamp -> Cache -> Maybe Cache
-expire1 now c =
-  ex =<< PSQ.minView c
+expires :: Timestamp -> Cache -> Maybe Cache
```

Comment:
I don't understand why `expires` returns `Maybe Cache` instead of `Cache` where `Cache empty xsz` can express an empty cache.

Reply:
Nothing if Cache does not change before and after expires, and Just if it does.